Parameterized tasks.

hjoliver opened the pull request:
Close #1937.
Close #1900.

Refactored graph-string parser.

New functionality:
 * parameterized tasks and families - i.e. auto-generate workflow without Jinja2 loops 
   * (this also enables "chunking" - splitting model runs into chunks (within a single cycle point) - without Jinja2)
   * parameter values are passed to task environments
   * any number of parameters (`foo<m>, bar<m,n>, baz<m,n,o>, qux<m,n,o,p>`, etc.)
   * works in cycling and one-off suites
 * ~~member-to-member family semantics, for same-size same-sort-order families~~ (removed)
 * syntax-driven line continuation

Works and the test battery passes, BUT:
 * I removed cylc-5 start-up and cold-start task back-compatibility and associated tests - the associated graph-parsing code looked like a nasty complication
   * (need to either restore this or officially obsolete cylc-5 syntax)

Still TODO:
 * ~~add more doc-strings and comments~~
 * ~~address a bunch of TODOs that I left in the code~~
 * ~~add tests for the new functionality~~
 * ~~add more error-detection for mis-use of the new syntax~~
 * ~~user guide documentation~~
 * ~~(address the back-compat issue mentioned above).~~

hjoliver commented:
## Examples

### Member-to-member triggers
```
[scheduling]
    [[dependencies]]
        graph = prep => GET_PETS => FEED_PETS => done
[runtime]
    [[GET_PETS]]
    [[get_cats, get_dogs, get_fish]]
        inherit = GET_PETS
    [[FEED_PETS]]
    [[feed_cats, feed_dogs, feed_fish]]
        inherit = FEED_PETS
```
![mem-to-mem](https://cloud.githubusercontent.com/assets/2362137/17163497/87f1e0f6-53af-11e6-9901-83bb86b10e1c.png)

matthewrmshin commented:
(Test battery passes in my environment with master merged in. The conflict is a tiny one in `cylc.scheduler`.)

hjoliver commented:
Branch rebased (I still need to address my own TODOs before review...)

hjoliver commented:
@matthewrmshin et. al. - this is now ready for review.   I'll quickly add documentation to the user guide once the notation and implementation is agreed.   Also can you discuss, do I need to add cylc-5 support back in as per remaining TODOs that I've left at the top of `config.py`:
```
# NOTE: OBSOLETED:
#  * OLD INITIAL TASKS (start-up and cold-start)
#  * IMPLICIT CYCLING
#  * ASYNC GRAPH IN CYCLING SUITE
#  * Ignored qualifiers on RHS, e.g. foo => FAM:succeed-all => bar
#    now must be "foo => FAM" and "FAM:succeed-all => bar".
#  * note cannot get-config retrieve an async graph anymore, it is moved to R1.
# TODO: CONSIDER OBSOLETING ALL cylc-5 syntax?
```
(As far as NIWA is concerned, cylc-5 is long dead and I'd rather not mess up the new graph parser with it!)

hjoliver commented:
Nice visualization of parameterized chunks:
![chunk1-fam](https://cloud.githubusercontent.com/assets/2362137/17725344/78996980-649e-11e6-980b-3f1a851277d3.png)
![chunk1-tsk](https://cloud.githubusercontent.com/assets/2362137/17725346/7c88ee30-649e-11e6-9713-55561eee3c6e.png)

matthewrmshin commented:
While I like the new *Paramterized tasks* syntax, I am not sure I understand the *Member-to-member triggers* syntax. Should the latter be an extension to the former? E.g. should we allow a value in the new *parameters* section to be either an integer (for a range) or a list of strings.

I guess we should recommend users to use more informative names as array indices (given that they get translated into environment variables)?

@arjclark please review 2 in parallel.

hjoliver commented:
> I guess we should recommend users to use more informative names as array indices (given that they get translated into environment variables)?

Yes (or up to the user).  For example purposes I just like to avoid hugely long task names.

matthewrmshin commented:
@hjoliver The test `./tests/authentication/10-remote-suite-passphrase-cache.t` is failing in my environment. Further investigation suggests that the test has caught a `cylc validate` failure if `graphing` is not installed on a platform. (In this case, `graphing` is not installed on the remote test host.)

```
Traceback (most recent call last):
  File "/home/matt/cylc.git/bin/cylc-validate", line 124, in <module>
    main()
  File "/home/matt/bin/cylc-validate", line 73, in main
    write_proc=not options.nowrite)
  File "/home/matt/cylc.git/lib/cylc/config.py", line 140, in get_inst
    vis_start_string, vis_stop_string, mem_log_func)
  File "/home/matt/cylc.git/lib/cylc/config.py", line 556, in __init__
    self.load_graph()
  File "/home/matt/cylc.git/lib/cylc/config.py", line 1773, in load_graph
    right, seq, suicide)
  File "/home/matt/cylc.git/lib/cylc/config.py", line 1286, in generate_edges
    graphing.edge(right, None, seq, suicide, conditional))
NameError: global name 'graphing' is not defined
```

hjoliver commented:
> should we allow a value in the new parameters section to be either an integer (for a range) or a list of strings.

That's not a bad idea.  Then we could use the string lists to generate the tasks under `[runtime]` by parameter expansion.  (Whereas I've hard-coded them in the example, and assumed you could use Jinja2 loops for the job).  I suppose it would make my new family notation `FAM => BAM` unnecessary, in that you could replace it with the parameterized member expression `f<i> => b<i>` (for `i = cat, dog, fish`, say).  But I still quite like the family notation as an even cleaner way of saying the same thing, where appropriate (including hard-wired similar task names like `get_ship_obs => proc_ship_obs` and `get_sonde_obs => proc_sonde_obs`, etc.).  (Note this is more like my original proposal too #1937 )

hjoliver commented:
Re failing test: it passes here; I'll address the problem tomorrow...

dpmatthews commented:
>  I suppose it would make my new family notation FAM => BAM unnecessary,

I'm not keen on this family notation. If I remember correctly, this notation used to mean `succeed-all`. We moved away from this, partly because it's too easy for users to write family triggers like this without really thinking about what they mean or without realising that it's a family rather than a task,

hjoliver commented:
All feedback addressed.  I suggest you hold off further review until I post an updated description later...

hjoliver commented:
So now, in response to feedback:
 * parameter values are now lists of string values.  The original integer values are now converted to strings for unified handling (works because no parameter value arithmetic is needed: an offset like `foo<obs-1>` can be interpreted as the previous item in the list of obs values).
 * to avoid ambiguity integer parameters (i.e. those that don't throw an exception on `int(value)`) expand to `_<name><value>`, e.g.
     * `foo<run>` expands to `foo_run0` and `foo_run1` for `run = 0, 1`
 * string list parameters are (a) unlikely to be ambiguous, and (b) longer than integers, so they expand to `_<value>` only, e.g.
     * `foo<size>` expands to `foo_small` and `foo_big` for `size = small, big`
 * as stated above, this makes member-to-member family triggers largely redundant, but not entirely redundant, so I've kept them but required them to be explicit:
     * `FAM:succeed-mem => BAM`
 * integer parameters are zero-indexed, because although one-indexed is probably more sensible for non-programmers, cylc users have to contend with zero-indexed Jinja2 anyway.
 * parameter values are now passed to tasks as `CYLC_TASK_PARAM_<NAME> = <VAL>` rather than just `<NAME> = <VAL>`

An example using,
```
[cylc]
   [[parameters]]
        run = 2
        pets = cats, dogs, fish
```
![string](https://cloud.githubusercontent.com/assets/2362137/17772179/dd86f964-6536-11e6-93a3-e11da68ac576.png)

The latest changes work but are not review-ready due to quick implementation today.  Let me know if you have any further critique, and I'll tidy up (and perhaps document) tomorrow.

matthewrmshin commented:
Hi @hjoliver I have run out of time today. I'll try to find time to comment on your latest proposal tonight or as early as possible tomorrow morning.

hjoliver commented:
> I have run out of time today...

No worries, there's no huge hurry - as long we can get this into a cylc-7 beta branch sometime in the next 4 weeks I'll be happy.

dpmatthews commented:
Some further thoughts:

 1. I'm not keen on the special treatment of single integer parameters (i.e. `run = 4` implies `run = 0,1,2,3,4`). Could we support `run = 0..4` instead? Then we could also have `run = 1..4`.

 2. Using the parameter name in the task name when using integer values may lead to problems choosing parameter names in large suites which use lots of parameters (and may result in very short parameter names or very long task names). Could we make the the task name a template instead? e.g. `foo_run<run>`, `foo_<size>`, `post_c<c>_r<r-1> => post_c<c>_r<r>`)? Would that work? Not quite as neat but more flexible and you get exactly the task names you want/expect.

 3. `FAM:succeed-mem` is better but I still don't think we need it. Using the parameterized tasks notation seems clearer to me and any validation errors should be obvious. What if the user has accidentally used the family in a normal task or in multiple parameterized tasks or what if the parameters used in the 2 families don't match? It could be difficult to validate or to understand the validation message.

(I haven't had a chance to discuss some of this with Matt so he may have different ideas)

matthewrmshin commented:
I think I broadly agree with @dpmatthews. A few things to add:

* We should ensure that an integer list such as `1..100` ends up with strings such as `*001` to `*100` in the task names, so they are easily sortable.
* On @dpmatthews's point 2, should we use Python template strings (or format strings)?
* It would be nice if we can use the same syntax (with braces, etc) as the final task names, but I guess we cannot easily do this (ambiguous syntax? shell meta-characters?)

hjoliver commented:
@dpmatthews @matthewrmshin - you guys drive a hard bargain :grimacing: 

 * Agreed on integer range syntax and padding.

 * I'm in two minds about dropping family member-to-member triggers.   The same result can be achieved with parameterized tasks, but in simple cases (such as I've shown) mem-to-mem triggers are probably easier for users (parameterization is a more "advanced" concept IMO).   Are you quite sure that supporting this would be a negative thing? 

 * I'm not convinced on your template syntax suggestions.   The purpose of our graph notation is to define dependencies in a way that shows suite structure as clearly as possible; and the purpose of this change is to avoid the need for the manual in-loop parameterization that wrecks this clarity.   We'll shoot ourselves in the foot if the new parameterization syntax itself ends up being unnecessarily complex and hard to read.   Consider replicating this small workflow over runs and members:
```
prep => sim => post = done
```
My syntax is very clear, in fact the base workflow is still visible at a glance:.  
```
prep => sim<r,m> => post<r,m> => done
```
Dave's suggestion muddies the waters significantly, by comparison:
```
prep => sim_r<r>_m<m> => post_r<r>_m<m> => done
```
And Matt's (a proper string template) will look like total gibberish to many users: 
```
prep => sim_r%(r)s_m%(m)s => post_r%(r)s_m%(m)s => done
```
Additionally, the second two suffer from use of the same symbol to represent both the fixed parameter name *and* the variable value to be substituted, e..g. `r` and `<r>` or `%(r)s`.  There are ways around this, but - it seems to me - at a cost of further syntactic complexity.

The only disadvantage of my syntax are that the underlying string template is fixed and implicit.  I would argue the implicit template hardly matters because you don't need to use the concrete task names anywhere in the suite definition (same notation `post<r,m>` in graph and runtime).   However, I guess we could allow explicit templates if needed, without reducing graph clarity, like this:
```
[[parameters]]
    r = 1..2
[[parameter templates]]
    r = _run%(r)s  
...
graph = prep => sim<r> => post<r> => done
```

hjoliver commented:
@matthewrmshin - I don't understand your final bullet point above - can you explain what you mean there?

matthewrmshin commented:
Hi @hjoliver sorry for being clearer in my comments. I am not really asking to use a Python template string in the graph. I am only asking that, when the task names are being expanded, can we use a Python template string? So we don't end up with hard coded logic such as:

> * `foo<run>` expands to `foo_run0` and `foo_run1` for `run = 0, 1`
> * `foo<size>` expands to `foo_small` and `foo_big` for `size = small, big`

My final bullet point is also related to the above. Can we support expanded task names such as `foo(0)`, `foo(small)`, etc?

hjoliver commented:
> So we don't end up with hard coded logic such as...

Perhaps I'm being thick (quite possible!) but I already convert the graph notation into a Python template string to use in the final expansion.  But do you mean something more than that?

> My final bullet point is also related to the above.

OK, understood on this one - yes, that's a good idea.

So, are you more or less happy with my notation in the graph then?

matthewrmshin commented:
Sorry for not being clear again. What I meant is that can the Python template string be configurable in `suite.rc`?

Yes. I am happy with your graph notation.

matthewrmshin commented:
(If there is a way to do `foo(run)` or `foo(size)` instead of `foo<run>` or `foo<size>` in graph, then I'll be happier, but that's probably asking too much.)

hjoliver commented:
More clarification required - you seem to want the same round parenthesis form for both the graph notation *and* the  expanded task names?? 

Notation:
> If there is a way to do foo(run) or foo(size) instead of foo<run> or foo<size> in graph ...

Expanded names:
> Can we support expanded task names such as `foo(0)`, `foo(small)`, etc?

Actually we don't currently support parentheses in task names.  I just tried it, and `foo(bar) => baz` gets interpreted as `foo & bar => baz` (there's a bug there somewhere...)

matthewrmshin commented:
No, we don't because parentheses are used in graph for expression precedence. While you are writing a new graph parser, I am just wondering aloud to see if parentheses can be used unambiguously in both contexts.

hjoliver commented:
The latest on this branch:

 * I've removed member-to-member triggers, due to popular lack of demand :grin: 

 * Integer parameters must now be expressed with range notation, e.g. `1..3` means `[1,2,3]`.

 * Integer parameter values are now zero-padded, but you don't have to use padding when singling out specific values, e.g. `foo<run=9>` may expand to `foo_run09`.

 * The default parameter expansion templates are still as described above (integer values are prefixed by name, non-integer string values aren't) but you can provide string templates to override the defaults:
```
[cylc]
     [[parameters]]
          run = 1..3
              [[[templates]]]
                   run = "-R%(run)s"    #  Override the default "_run%(run)s",
[scheduling]
       [[dependencies]]
             graph = "foo<run> => bar<run>"
```

  * Fixed the graphing import problem above https://github.com/cylc/cylc/pull/1953#issuecomment-240376765

Finally, I considered changing the notation from angle to round brackets, but (a) it looks difficult to do, and (b) I prefer the angle brackets as they make the parameterization stand out more obviously.

matthewrmshin commented:
This is looking good. Some minor but hopefully final comments...
* Would it be better if we have `[[paremeter-templates]]` (at the same level as `[[parameters]]` (instead of `[[[templates]]]`, just in case some users decide to use `templates` as the name of a parameter?
* I don't actually mind angle brackets, apart from their tendency to upset HTML rendering every now and then.

hjoliver commented:
@matthewrmshin - 
 * `[[parameter templates]]` is what I did initially before changing to use a sub-section of `[[parameters]]`.  But I think you're right, I'll switch back.
 * I actually didn't think of the HTML implications of angle brackets.  Really that's a problem to be solved by systems that want to display arbitrary pre-formatted text in a web page, I suppose ... but I don't want to make life difficult for `rose bush` in particular - will that be much of a problem?

matthewrmshin commented:
Rose Bush should be fine.

hjoliver commented:
* parameter template config moved up one level, as discussed
 * documented parameterized tasks in a new section of the User Guide (includes discussion of exactly how this relates to dynamic cycling)

@matthewrmshin - this can be reviewed now.   We just need to decide whether this branch becomes the basis for cylc-7 (along with HTTP comms) or I restore cylc-5 start-up and cold-start functionality for earlier merge to master (I'll only do the latter if it's easy...)

hjoliver commented:
The user guide Jinja2 section needs rewriting now (simplifying really) because it's largely focused on generating tasks in loops, which is made more or less redundant by this branch.  But it's not technically incorrect as-is, so the rewrite is not too urgent.

hjoliver commented:
(Branch got the Travis CI tick - it's a miracle :grin:)

arjclark commented on doc/cug.tex:
```diff
@@ -1,4 +1,4 @@
-\lstset{language=transcript}
+\lstset{language=transcript} 
```
Random extra whitespace?

hjoliver commented:
cylc-5 back-compatibility and associated tests restored (which was not much fun :grimacing:)

hjoliver commented:
### performance

There's a small performance hit on this branch, on average, as might be expected (it does more, and is less "inlined" than on master).  However, there's not much in it.  For repeated validations of `dev/busy` on my laptop VM with 5000 tasks per cycle, this branch actually wins sometimes:
```
$ cylc list bs | wc -l
5000
```
This branch:
```
$ git checkout 1937.parameterized-tasks

$ time cylc val bs
Valid for cylc-6.10.2-208-g9e04e

real	0m11.204s
user	0m10.053s
sys	0m1.131s

$ time cylc val bs
Valid for cylc-6.10.2-208-g9e04e

real	0m10.460s
user	0m10.215s
sys	0m0.482s
```
Master:
```
$ git checkout master

$ time cylc val bs
Valid for cylc-6.10.2-178-ga64c7

real	0m10.034s
user	0m9.744s
sys	0m0.474s

$ time cylc val bs
Valid for cylc-6.10.2-178-ga64c7

real	0m10.715s
user	0m9.662s
sys	0m1.023s
```

So, given that backward compatibility has been restored, the test battery passes, and performance is pretty good, I propose that this goes into next release (ready for the tutorial VMs...)

hjoliver commented:
(TCI failed due to our old friend `./tests/pre-initial/11-warm-insert-stall.t` and `./tests/ext-trigger/00-satellite.t` which is also flaky on TCI, nothing specific to this branch)

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+
+\subsubsection{Parameter Expansion}
+
+Parameters values can be lists of strings, or integer ranges (both bounds included):
```
Parameter~~s~~

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+\begin{lstlisting}
+    proc_ship, proc_buoy, proc_plane
+\end{lstlisting}
+For integer values the parameter name appears in the expanded task name, but
```
Not sure I understand what's being said here and the impact of this limitation?

hjoliver replied:
It's not a limitation, it's a feature.   If you have parameters `m` and `n` (say) with integer values, the values alone don't tell you which is which, so by default we expand `foo<m,n>` to `foo_m1_n2` (etc.) rather than  `foo_1_2` (does the `2` come from `m` or `n`?). 

I will make this clearer though.

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+\end{lstlisting}
+For integer values the parameter name appears in the expanded task name, but
+for string values it doesn't.  These default expansion templates  can be
+overridden if necesasry:
```
necessary

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+\subsubsection{Passing Parameter Values To Tasks}
+
+Parameter values are passed as environment variables to tasks generated by
+parameter expansion.  For the example above, task \lstinline=model_run2_ship=
```
Extra space before "For"

hjoliver replied:
Ha, that's a habit I learned in typing classes in high school (on *typewriters*!) .  The modern rule does seem to be one space after a period.  LaTeX ignores extra spaces in any case (but, interestingly, and it apparently typesets a longer space after a period, by default).

benfitzpatrick replied:
It's also in PEP 8! https://www.python.org/dev/peps/pep-0008/#comments

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+\subsubsection{Parameterized Tasks In The Graph}
+
+Using a different number of parameters on each side of an arrow in the graph
+will branch or join a workflow.  For example this,
```
this:

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+\begin{lstlisting}
+    graph = "prep => model<run>"  # for run = 1, 2
+\end{lstlisting}
+expands to this,
```
this:

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+    prep => init_run2 => model_run2_cat => post_run2_cat => wrap_run2 => done
+            init_run2 => model_run2_dog => post_run2_dog => wrap_run2
+\end{lstlisting}
+This is visualized by \lstinline=cylc graph= in Figure~\ref{fig-params-1}.
```
Think I've lost track of what "This" is at this point...

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+    [[model<run>]]
+        # config for all "model" runs...
+    [[model<run=1>
+        # special config for the first model run (if any)...
```
For my peace of mind, I gues model<run=1> gets all the values in model<run> and then just overrides whatever special item there is defined under it (inheritance style)?

hjoliver replied:
It's override by repetition (not inheritance): the first heading configures all `model<run>` including for `run=1`, the second one just (re)configures `model<run=1>`.

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+\end{lstlisting}
+
+As described in Section~\ref{FamilyTriggers} family names can be used to
+trigger all members at once,
```
once:

hjoliver replied:
I thought it was OK to use a comma or a colon in this context.  The user guide is probably riddled with it.  I'll see if I can find any guidance on this...

arjclark replied:
I only picked up on this one really as its inconsistent with what follows where you are using colons in the same paragraph.

hjoliver replied:
I was thinking of it as variation, using both tools in the box - but you're right, it is inconsistent!

hjoliver replied:
Alright, doing a mass conversion to colon...

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+\end{lstlisting}
+and this graph:
+\begin{lstlisting}
+    graph = "proc<size-1> => proc<size>"  # for size = small, big, huge
```
How would `proc<size-1> => proc<size> => elephant` be handled?

hjoliver replied:
The same as in the example, but additionally `elephant` will trigger off all `proc<size>`.

matthewrmshin commented:
I am unable to get `./tests/validate/40-fail-suicide-left.t` to pass in my environment. Keep getting:

```
(SHELL)$ CYLC_CONF_PATH= cylc test-battery -v ./tests/validate/40-fail-suicide-left.t

ok 1 - 40-fail-suicide-left
--- 40-fail-suicide-left.stderr	2016-09-07 12:18:13.725619381 +0100
+++ -	2016-09-07 12:18:13.742289106 +0100
@@ -1,3 +1 @@
-/home/matt/cylc.git/lib/cylc/graph_parser.py:32: DeprecationWarning: BaseException.message has been deprecated as of Python 2.6
-  return self.message.replace('=>', ' => ')
 ERROR, suicide markers must be on the right of a trigger: !dont-kill-me
not ok 2 - 40-fail-suicide-left.stderr-cmp-ok
```

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+    get_plane => proc_plane
+\end{lstlisting}
+
+\subsubsection{Static Chunking vs Dynamic Cycling}
```
I guess I find this section a bit misleading tbh. Going in to it I was expecting some new form of dynamic cycling but what we really have is an R1 with a list of N entries where the task has to manipulate the nth parameter to generate a datetime. (I was also half expecting this to be dynamic sets of values e.g. at one cycle `<param>` evals as 1..3 whereas at another it evals as 1..10)

hjoliver replied:
(email follow-up coming...)

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+Parameter expansion can be used instead of using dynamic cycling to generate
+repeated task runs.  The resulting series of tasks may then be referred to as
+steps or ``chunks''. The fundamental difference between these approaches is
+that {\em parameter expansion is static} (parameters are expanded at start-up,
```
shouldn't \em be on the left of the { ?

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+time, potentially indefinitely.
+
+Here's a dynamic cycling workflow of two-monthly model runs for one year, with
+previous-instance model dependence (e.g.\ for model restart files):
```
is the "\ " to force a space there?

hjoliver replied:
Tells LaTeX that this period/full-stop does not signify the end of a sentence, which it typesets differently - see the two-spaces discussion above.

benfitzpatrick commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+repeated task runs.  The resulting series of tasks may then be referred to as
+steps or ``chunks''. The fundamental difference between these approaches is
+that {\em parameter expansion is static} (parameters are expanded at start-up,
+when the suite is parsed), whereas dynamic cycling extends a workflow at run
```
I'm not sure about 'Dynamic Cycling' - it brings up thoughts of asynchronous spawning and things like that.
Maybe 'indefinitely extended' or 'unbounded' cycling? Or just 'cycling'?

hjoliver replied:
(email follow-up coming...)

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+        cycle point is far in the future, or if there is no final point at all.
+    \item Dynamic cycling is also better for suite visualization and monitoring.
+        \begin{myitemize}
+            \item Every task in the entire static workflow is necessarily
```
I guess, N.B this is actually a particularly bad thing in graph view as it'll eat all your memory...

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+    \caption[Static chunked (top) and dynamic cycled (bottom) versions of the
+    same workflow.]{\scriptsize Static chunked and dynamic cycled versions of
+    the same workflow.}
+    \label{fig-eg2}
```
Not seeing the images for this in my html copy of the docs.

hjoliver replied:
Good point, hadn't checked the HTML - will fix tomorrow NZ.

hjoliver replied:
Fixed.

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+                the dynamic case you can visualize as little or as much of it as
+                you like, and at run time you'll only see cycle points that are
+                currently active.  The two workflows are shown together in
+                Figure~\ref{fig-eg2}.
```
So, I guess I'm not clear how this is working differently. Is it that in the R1 "static" case everything is hard coded, so on evaluation for visualisation everything is drawn, whereas in the "dynamic" case the items are evaluated as you go, resulting in it only visualising where its up to in the evaluation chain?

hjoliver replied:
Yep.   Suggestions welcome on explaining more clearly...

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+                Figure~\ref{fig-eg2}.
+            \item In the dynamic workflow successive instances of the same task
+                have the same name and are labelled with their proper date-time
+                cycle points.  In the static workflow successive instances of
```
- since it is an expanded out parameter

arjclark commented on doc/cug.tex:
```diff
@@ -4551,6 +4489,380 @@ \subsubsection{Collapsible Families In Suite Graphs}
+initial cycle point.  In the future we plan to provide this kind of
+simplification at the start of any date-time sequence, not just the suite
+initial point.  It will be optional though, because delayed-start cycling tasks
+sometimes need to be triggered off earlier tasks.
```
I guess I'm not sure what the equivalent suites are here that this applies to. From the earlier context I thought we were comparing this sort of thing:
```
[scheduling]
[[dependencies]]
[[[R/P1D]]]
  graph = model[-P1D] => model
```
to this
```
[scheduling]
[[dependencies]]
graph = model<chunk-1> => model<chunk>
```

So, I'm not clear where the gain is as the two are pretty much the same with magic to cut off the first invalid/non-existent referenced entry?

hjoliver replied:
You're right for a single sequence, but consider replicating that over `run = 1, 2` in the same suite:
```
     graph = model<run, chunk-1> => model<run, chunk>
```
in which `model<run=1>` is initialized at date-time T1, but `model<run=2>` is initialized at T2 > T1.  In the "dynamic cycling" version of this, only T1 can be the suite initial cycle point, so T2 does not get the start-up magic.

hjoliver replied:
(this is relevant to suites in the scheduler comparison report)

arjclark commented:
`./tests/validate/40-fail-suicide-left.t` also failing to pass in my environment

hjoliver commented:
> ./tests/validate/40-fail-suicide-left.t also failing to pass in my environment

Huh, it passes here; I don't get the deprecation message (as reported by @matthewrmshin) from Python 2.7.5.

hjoliver commented:
New draft CUG section on "cycling" added.  I still need to address the remaining feedback above.

hjoliver commented:
New draft section revised.  Will address remaining feedback tomorrow (and review the latest tutorial draft)

matthewrmshin commented:
Tests OK in my environment.

hjoliver commented:
OK:
 * Revised Section 5.  Formerly *On the meaning of "cycle" and "cycle point" in cylc*, now it is *Cycling and Cycle Points in cylc*, with short subsections on terminology and "ways of cycling" (dynamic vs static cycling)
 * Revised the Parameterized task section in light of feedback.

I'm personally pretty happy with the documentation now.  If you disagree only mildly, maybe consider improvements via later PRs, depending on current levels of urgency (when do you want to get the tutorial VM out?)

arjclark commented:
@matthewrmshin - review 2 ok. In the interests of time, agree with @hjoliver that documentation can be refined at a later date (I still don't like the way "dynamic" is being used here but it's not a blocker on the otherwise good work).

As noted by @hjoliver above, the only thing I've noted is a slight slowdown in validation times with master (at time of investigating) giving a validation time of:
```
time CYLC_CONF_PATH= cylc validate .
Valid for cylc-6.10.2-191-gadd7f

real	0m29.154s
user	0m27.716s
sys	0m1.001s
```
vs this branch of:
```
time CYLC_CONF_PATH= cylc validate .
Valid for cylc-6.10.2-217-g09085

real	0m37.087s
user	0m34.902s
sys	0m0.901s
```
This is testing with a suite based on the horrible numbers of inter-family triggers as:
```
#!jinja2
{% set NUM_MEMBERS_FAM1=200 %}
{% set NUM_MEMBERS_FAM2=200 %}
[scheduling]
[[dependencies]]
graph = FAM1:succeed-all => FAM2
[runtime]
[[root]]
script = true
[[FAM1]]
[[FAM2]]
{% for member in range(1,NUM_MEMBERS_FAM1) %}
[[fam1_member_{{member}}]]
inherit = FAM1
{% endfor %}
{% for member in range(1,NUM_MEMBERS_FAM2) %}
[[fam2_member_{{member}}]]
inherit = FAM2
{% endfor %}
```
I don't think this is a showstopper and not an unexpected slowdown given the increased work being done, just noted here for completeness.

@matthewrmshin - over to you to press the button when you're happy for this to go in.

hjoliver commented:
Woo-hoo, it's merged at last!

For the record, I'm seeing these timings on the suite with "horrible numbers of inter-family triggers"

 * cylc-6.10.2 : 16.5s
 * cylc-master (with this now merged in): 20.3s

This isn't too much of a hit, but I'll do some more profiling in due course...
